integration/insert_compat_test.go

```diff
@@ -0,0 +1,143 @@
+			insert:     bson.D{},
+			resultType: nonEmptyResult,
+		},
+		"InsertIDArray": {
+			insert:     bson.D{{"_id", bson.A{"foo", "bar"}}},
+			resultType: emptyResult,
+		},
+		"InsertIDRegex": {
+			insert:     bson.D{{"_id", "^regex$"}},
+			resultType: emptyResult,
+		},
+	}
+
+	testInsertCompat(t, testCases)
+}
+
+type insertCompatTestCase struct {
+	insert        bson.D                   // required
+	resultType    compatTestCaseResultType // defaults to nonEmptyResult
+	skip          string                   // skips test if non-empty
+	skipForTigris string                   // skips test for Tigris if non-empty
+}
+
+// testInsertCompat tests insert compatibility test cases.
+func testInsertCompat(t *testing.T, testCases map[string]insertCompatTestCase) {
+	t.Helper()
+
+	for name, tc := range testCases {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Helper()
+
+			if tc.skip != "" {
+				t.Skip(tc.skip)
+			}
+			if tc.skipForTigris != "" {
+				setup.SkipForTigrisWithReason(t, tc.skipForTigris)
+			}
+
+			t.Parallel()
+
+			ctx, targetCollections, compatCollections := setup.SetupCompat(t)
+
+			insert := tc.insert
+			require.NotNil(t, insert)
+
+			var nonEmptyResults bool
+			for i := range targetCollections {
+				targetCollection := targetCollections[i]
+				compatCollection := compatCollections[i]
+				t.Run(targetCollection.Name(), func(t *testing.T) {
+					t.Helper()
+
+					var targetInsertRes, compatInsertRes *mongo.InsertOneResult
+					var targetErr, compatErr error
+
+					targetInsertRes, targetErr = targetCollection.InsertOne(ctx, insert)
+					compatInsertRes, compatErr = compatCollection.InsertOne(ctx, insert)
+
+					if targetErr != nil {
+						t.Logf("Target error: %v", targetErr)
+						targetErr = UnsetRaw(t, targetErr)
+						compatErr = UnsetRaw(t, compatErr)
+
+						// Skip inserts that could not be performed due to Tigris schema validation.
+						var e mongo.CommandError
+						if errors.As(targetErr, &e) && e.Name == "DocumentValidationFailure" {
+							if e.HasErrorCodeWithMessage(121, "json schema validation failed for field") {
+								setup.SkipForTigrisWithReason(t, targetErr.Error())
+							}
+						}
+
+						assert.Equal(t, compatErr, targetErr)
+					} else {
+						require.NoError(t, compatErr, "compat error; target returned no error")
+					}
+
+					targetID, _ := pointer.Get(targetInsertRes).InsertedID.(primitive.ObjectID)
+					compatID, _ := pointer.Get(compatInsertRes).InsertedID.(primitive.ObjectID)
+
+					if !(targetID.IsZero() && compatID.IsZero()) {
+						nonEmptyResults = true
+					}
+
+					var targetFindRes, compatFindRes bson.D
+					targetCur, err := targetCollection.Find(ctx, bson.D{{}})
+					require.NoError(t, err)
+					targetCur.Decode(&targetFindRes)
+
+					compatCur, err := compatCollection.Find(ctx, bson.D{{}})
+					require.NoError(t, err)
+					compatCur.Decode(&targetFindRes)
+
+					AssertEqualDocuments(t, compatFindRes, targetFindRes)
```

Comment:
Should we find by `_id` to compare inserted documents here? (The idea is to check that what we inserted is really stored equally).

Reply:
I think that checking whole collection content is much more full-bodied. What if the insert function misbehaves and creates additional entry in collection? Checking only for specific `_id` would not cover this case